### ISSUE TITLE
App Flavor not available after publishing in Play store via `flutter build appbundle`. The env variable is not included.

### ISSUE THREAD

**Issue by JiangHongTiao** [flutter/flutter#156661]
### Steps to reproduce

When I build my application and run it in emulator, I can clearly work with flavor or environmental variables. However when I build aab android package and upload it into GPC, mobile app doesn't know flavor, nor environmental variables. 

1. Build app as: `flutter build appbundle --release --flavor internal --dart-define="APP_FLAVOR=internal"`
2. When the app is installed through Android Play Store, ` String.fromEnvironment('FLUTTER_APP_FLAVOR')} / String.fromEnvironment('APP_FLAVOR')` are empty
3. When I build APK with command: ` flutter build apk --release --flavor internal --dart-define="APP_FLAVOR=internal"` and upload it into emulator, flavor is correctly displayed. 

### Expected results

Both env. variables `FLUTTER_APP_FLAVOR` and `APP_FLAVOR` contain string `internal`

### Actual results

Both env. variables `FLUTTER_APP_FLAVOR` and `APP_FLAVOR` are empty.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Text("appFlavor: $appFlavor", textDirection: TextDirection.ltr),
        Text("FLUTTER_APP_FLAVOR: " + const String.fromEnvironment('FLUTTER_APP_FLAVOR'), textDirection: TextDirection.ltr),
        Text("APP_FLAVOR: " + const String.fromEnvironment('APP_FLAVOR'), textDirection: TextDirection.ltr),
      ],
    );
  }
}

```

Define new Flavor in Android project ./Android/app/build.gradle:
```gradle
android {
    flavorDimensions "version"
    productFlavors {
        internal {
            dimension "version"
            versionNameSuffix "-internal"
        }
    }
}
```

```bash
flutter run --flavor internal --dart-define="APP_FLAVOR=internal"
```
</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

Emulator
![image](https://github.com/user-attachments/assets/fc665c58-746b-4637-8f0b-32fbe44f1bd3)

Mobile device (Pixel 7); App installed through Google Play Store:
![image](https://github.com/user-attachments/assets/ff04d761-642c-4a88-9493-d78383115bf4)

</details>

### Logs

-

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
bug % flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.24.3, on macOS 14.7 23H124 darwin-arm64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2024.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2024.2.1)
[✓] Connected device (5 available)
[✓] Network resources

• No issues found!
```
</details>

**Comment by darshankawar** [comment]
Thanks for the report @JiangHongTiao 
Can you take a look at [this](https://github.com/flutter/flutter/issues/142976) issue along with underlying comments specific to Android to see if they help in your case or not ?

https://github.com/flutter/flutter/issues/142976#issuecomment-1933401058
https://github.com/flutter/flutter/issues/142976#issuecomment-1949602247 (Android part)

**Comment by JiangHongTiao** [comment]
Thanks @darshankawar for quick response. 
However the result is the same regardless I use `--dart-define-from-file` or `--dart-define`. 
`flutter build appbundle` seems not to include those env variables.

**Comment by darshankawar** [comment]
> `flutter build appbundle` seems not to include those env variables.

Maybe it is similar to https://github.com/flutter/flutter/issues/56792 that you can check with underlying comments.

**Comment by JiangHongTiao** [comment]
Hi @darshankawar ,
I'm not sure if I understand comments in that thread. The thread seems to have no resolution, they only speak about proposals or  `--dart-define-from-file` or `--dart-define`. 
What was the solution you were considering could work?

**Comment by darshankawar** [comment]
Sorry if it wasn't clear. I actually wanted to share that issue that discussed a proposal to allow `--dart-define` from environmental var but per team member's [comment](https://github.com/flutter/flutter/issues/56792#issuecomment-626255953), it probably isn't implemented or supported by the tool and solution was as described in it:

`write a script to populate a flutter run command with all the configuration that you need.`

**Comment by JiangHongTiao** [comment]
Thanks @darshankawar , but it doesn't really solve my problem. I have only one variable I want to have in the build, but it's not simply there. I need that one of parameters `--dart-define-from-file` or `--dart-define` works, but it does not.

**Comment by darshankawar** [comment]
Thanks for the update @JiangHongTiao 
I followed the code sample and steps to replicate and by running the said command in android emulator with which the string is displayed properly as below:

<img width="432" alt="Screenshot 2024-10-17 at 11 20 05 AM" src="https://github.com/user-attachments/assets/12ee1fc3-1aa2-4890-9637-e80be04fc287">

I created the appbundle but I think in order to verify it, it need to be deployed in playstore for which I need to set it up. Meanwhile, the issue is, upon running below command:

`flutter build appbundle --release --flavor internal --dart-define="APP_FLAVOR=internal"`, it isn't reading the `app_flavor` values only after deployed to playstore, is this a correct issue behavior but runs properly locally and upon building apk ?

**Comment by JiangHongTiao** [comment]
Yes @darshankawar , that's correct. I can only reproduce it when it's installed from Google Play Store. 
What I have tried so far is:

Build APK
1. Install APK into emulator ✅
2. Install APK into Physical device ✅

Build `AAB` and convert it into `APK` with `bundletool` as:
```bash
brew install bundletool
bundletool build-apks --bundle=app-release.aab --output=app-release.apks
bundletool install-apks --apks=app-release.apks
```
1. Install converted APK into emulator ✅
2. Install converted APK into Physical device ✅

However uploading it into Google Play Store and installing it in the phone has no `flavor` information at runtime.

**Comment by darshankawar** [comment]
Thanks for the update. Keeping the issue open for team's tracking.

**Comment by reidbaker** [comment]
@JiangHongTiao Closing this as working as intended. Environment variables are local to the machine and when the play store builds you app it has no way to know what environment variables to set. Here is the android documentation for flavors as shipping a custom flavor is something that is supported. https://developer.android.com/build/build-variants#change-app-id I would recommend you modify your code that evaluates environment variables to assume the value you expect when it is not set and follow the linked instructions for shipping a custom build variant.

**Comment by andrewkolos** [comment]
Hello, if the [`appFlavor`](https://api.flutter.dev/flutter/services/appFlavor-constant.html) API from `package:flutter/services.dart` is broken upon publishing, then there is a legitimate bug here. @JiangHongTiao can you confirm for me that this is the case?

**Comment by JiangHongTiao** [comment]
Hi @andrewkolos ,
thanks for reopening. I was trying to reproduce it on a clean project and it seems to be working for me. However on my regular project I do have a problem with it. I'm not sure why. I've also tried different combinations with `static` and `const` keywords, but on a clean project it works. 
For now I will close it and if I find more evidence why it's happening, I will reopen it. Thanks all for your time.
